resgroup: fix a race condition issue.

pivotal-ning-yu opened the pull request:
Updates to memSharedUsage need to check memSharedGranted, while updates
to memSharedGranted need to check memSharedUsage, so there can be a race
condition on concurrent updates to memSharedUsage and memSharedGranted.

To fix this issue we protect these updates with a spinlock.

pivotal-ning-yu commented:
BTW, no test case is included in this PR as the issue doesn't lead to observable error and the fail rate is so low that we can only trigger it with the help of gdb so far.

hlinnaka commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -747,32 +753,26 @@ ResGroupReserveMemory(int32 memoryChunks, int32 overuseChunks, bool *waiverUsed)
-	/* reserve from slot memory */
-	slotMemUsage = pg_atomic_add_fetch_u32((pg_atomic_uint32 *)&slot->memUsage, memoryChunks);
-	slotMemSharedNeeded = slotMemUsage - slot->memQuota;
+	S_LOCK(&group->lock);
```
Please use the hardware-independent macros SpinLockInit, SpinLockAcquire, SpinLockRelease, instead of these low-level S_* macros. See comments at top of spin.h and s_lock.h.

xiong-gang commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -1422,20 +1436,22 @@ groupApplyMemCaps(ResGroupData *group, const ResGroupCaps *caps)
 	if (memStocksToFree > 0)
-	{
-		returnChunksToPool(group->groupId, memStocksToFree);
 		group->memQuotaGranted -= memStocksToFree;
```
Why it's not in spin lock?

xiong-gang commented:
LGTM except the spinlock API

pivotal-ning-yu commented:
@hlinnaka by the way, I was told that there might be some concern on the use of spinlock, could I know your thoughts on this please? Thanks in advance.

pivotal-ning-yu commented:
Some offline comments from @pengzhout and @xiong-gang, while holding the spinlock a failed assertion may cause the spinlock blocks other processes. So it's better to postpone these assertions until spinlock is released.

pivotal-ning-yu commented:
A commit is pushed to move the assertions after the release of the spinlock.

pengzhout commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -2249,9 +2275,16 @@ ResGroupHashRemove(Oid groupId)
+	if (toFree > 0)
+		returnChunksToPool(groupId, toFree);
+
 	group->groupId = InvalidOid;
```
should group->groupId be set to InvalidOid before setting of group->memQuotaGranted and memSharedGranted?

pivotal-ning-yu replied:
Good catch.

Once groupId is set to InvalidOid it's considered as a dropped resgroup, so in palloc() or pfree() the memory accounting information won't be updated or checked.

So yes I think we should mark it as dropped before touching the accounting information.

xiong-gang commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -1700,22 +1743,33 @@ groupReleaseMemQuota(ResGroupData *group, ResGroupSlotData *slot)
 		group->memQuotaGranted -= memQuotaToFree; 
-	}
+
+	SpinLockAcquire(&group->lock);
```
After use spinlock, we can carefully remove some of the lwlock
For example, in function SwitchResGroupOnSegment(), it acquires lwlock before call groupReleaseMemQuota(), i think we can remove it

pivotal-ning-yu commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -728,51 +734,88 @@ ResGroupReserveMemory(int32 memoryChunks, int32 overuseChunks, bool *waiverUsed)
+	 */
+	SpinLockAcquire(&group->lock);
+
 	if (group->groupId != self->groupId)
```
As we are also refactoring the resgroup source code in parallel, the code base is changing in a fast pace. While I trying to resolve the conflicts with current origin/master I encountered one problem.

The policy of current refactor is to abstract most reusable logic into small helper functions, many assertions are put into these helper functions so we will have stricter checks on the running status.

However as we now introduce spinlock to protect some updates, we don't want assertion to be put in spinlock blocks.

For example this check `group->groupId != self->groupId` is now replaced with a function `selfIsAssignedDroppedGroup()` in master, however as it contains assertions we can not use it while holding the spinlock, so we have to keep using this `!=` check which is less meaningful than the helper function.

Generally speaking assertion handling in spinlock blocks makes the code difficulty to understand, we have made some attempts in this PR but as a result I think the code readability is down as a cost.

I think we need to discuss the policy on this.

pivotal-ning-yu commented:
Rebased with origin/master as the code was refactored in master.

Also made some updates per our online & offline discussion:

* don't attempt to move assertions outside of spinlock for now;
* replace some uses of LWLock with the spinlock;
* added some debugging support for the spinlock;

xiong-gang commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -1521,14 +1703,14 @@ getChunksFromPool(Oid groupId, int32 chunks)
+	LOG_RESGROUP_DEBUG(LOG, "Free %u to pool(%u) chunks from group %d",
+					   chunks, pResGroupControl->freeChunks, group->groupId);
+
 	pResGroupControl->freeChunks += chunks;
```
This could be concurrently updated by different group, so it should be atomic

pivotal-ning-yu replied:
Ah, good catch, I'll update this as well as `getChunksFromPool()`.

xiong-gang commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -1052,84 +1138,172 @@ ResGroupCreate(Oid groupId, const ResGroupCaps *caps)
+
+		if (overuse > threshold)
+		{
+			/* The overuse is too large, give up */
```
spin lock is not released?

pivotal-ning-yu replied:
Oops, will correct this.

pivotal-ning-yu commented:
Per our offline discussion we will not apply this patch for now. We might have to find out a better solution. Closed, thanks for all the review.
